master/internal/trial.go

```diff
@@ -394,6 +406,59 @@ func (t *trial) Receive(ctx *actor.Context) error {
 	return nil
 }
 
+func (t *trial) registerRendezvousWatcher(ctx *actor.Context, msg trialWatchRendezvousInfoReq) {
+	// Validate this watch request is unique and not stale.
+	if _, ok := t.containerRanks[msg.containerID]; !ok {
+		ctx.Respond(apiutils.AsErrBadRequest(
+			"rendezvous request from stale container: %s", msg.containerID,
+		))
+		return
+	} else if _, ok := t.rendezvousWatchers[msg.containerID]; ok {
+		ctx.Respond(apiutils.AsErrBadRequest(
+			"rendezvous request from already connected container: %s", msg.containerID,
+		))
+		return
+	}
+
+	// Register and respond with the watcher channel. Channel is
+	// size 1 since rendezvous info will only ever be sent once.
+	w := make(chan interface{}, 1)
+	t.rendezvousWatchers[msg.containerID] = w
+	ctx.Respond(trialWatchRendezvousInfoResp{addresses: w})
+
+	// If we're not all ready, send the all ready timeout notification and return.
+	t.lastContainerConnectedTime = time.Now()
+	if !t.allReadyV2() {
+		actors.NotifyAfter(ctx, allReadyTimeoutPeriod, allReadyTimeout{runID: t.RunID})
+		return
+	}
+
+	t.pushRendezvouzV2()
+}
+
+func (t *trial) registerPreemptionWatcher(ctx *actor.Context, msg trialWatchPreemptionReq) {
+	if len(t.allocations) == 0 {
+		ctx.Respond(apiutils.AsErrBadRequest(
+			"no preemption status available for unallocated trials",
+		))
+		return
+	}
+
+	// Register and respond with the watcher channel. Size 2; at most 2 messages
+	// can be sent and we don't want to block or lose them.
+	w := make(chan bool, 2)
+	t.preemptionWatchers[msg.id] = w
+	ctx.Respond(trialWatchPreemptionResp{signal: w})
+
+	if t.PendingGracefulTermination {
```

Comment:
Why did you change this?

Reply:
I felt `len(allocations) == 0` was a more correct check than `trial_state != ACTIVE`.

Reply:
ended up adding cleanup in `t.terminated()` too, so the lifetime of a rendezvous or preemption watcher should only be from one `sproto.ResourcesAllocated` to the `t.terminated(ctx)` call to clean them up. This lifetime will become cleaner in an upcoming refactor.